astropy/table/table.py

```diff
@@ -355,8 +355,10 @@ def as_array(self, keep_byteorder=False, names=None):
         table_array : np.ndarray (unmasked) or np.ma.MaskedArray (masked)
             Copy of table as a numpy structured array
         """
+        empty_init = ma.empty if self.masked else np.empty
+
         if len(self.columns) == 0:
-            return None
+            return empty_init(0)
```

Comment:
Did it not work to just remove the lines? I thought this would end up doing `empty_init([], dtype=[])` which gives a nice empty array...

Reply:
I thought it would be better to return an appropriate "kind" of array (i.e., masked) compatible with the "kind" of the table.

Reply:
Also stacking a series of empty arrays (a silly example: `np.hstack([Table().as_array(), Table().as_array()])`) would work while with `np.void` type arrays it would crash (I think).

Reply:
No, that works fine. Could you try if anything more fails if you just remove the `if` clause? (Sorry, would do it myself but have to run...)

Reply:
@mhvk All tests pass either way: with or without `if` clause.

Reply:
I'd suggest to just remove the `if` clause, then.

Reply:
Are you sure? As I said above, it _feels_ "better" to "match" the kind of the returned array to the table (masked or not). For example, as is,
```python
In [33]: t = Table(data=None, masked=True)                                                                                      

In [34]: t                                                                                                                      
Out[34]: 
<Table masked=True length=0>
<No columns>

In [35]: t.as_array()                                                                                                           
Out[35]: 
masked_array(data=[],
             mask=False,
       fill_value=1e+20,
            dtype=float64)
```

Reply:
But I do not have strong feelings here. Just please confirm your decision.

Reply:
But if you remove the `if` clause, you would get the same Masked Array, except with a strange dtype, wouldn't you?

Reply:
Just tested and yes, it looks like it.

Reply:
Great, looks like we're all set here, then!

Reply:
Wait. Not ready.